lib/actions/deploy.coffee

```diff
@@ -0,0 +1,168 @@
+getBuilderPushEndpoint = (baseUrl, owner, app) ->
+	"https://builder.#{baseUrl}/v1/push?owner=#{owner}&app=#{app}"
+
+parseInput = (params, options) ->
+	if not params.appName?
+		throw new Error('Need an application to deploy to!')
+	appName = params.appName
+	image = undefined
+	if params.image?
+		if (options.build or options.context?)
+			throw new Error('Build and context parameters are not applicable when specifying an image')
+		options.build = false
+		image = params.image
+	else if options.build
+		context = options.context || '.'
+	else
+		throw new Error('Need either an image or a build flag!')
+
+	return [appName, options.build, context, image]
+
+pushProgress = (imageSize, request, timeout = 250) ->
+	process.stdout.write('Initialising...')
+	progressReporter = setInterval ->
+		sent = request.req.connection._bytesDispatched
+		percent = (sent / imageSize) * 100
+		if percent >= 100
+			clearInterval(progressReporter)
+			percent = 100
+		process.stdout.clearLine()
+		process.stdout.cursorTo(0)
+		process.stdout.write("Uploaded #{percent.toFixed(1)}%")
+		console.log() if percent == 100
+	, timeout
+
+# A regex for parsing;
+# * A hostname e.g. registry.resin.io
+# * An appname, which consists of 1 or more alphanumeric chars
+# * A commit hash (40 alphanumeric characters)
+resinImageRegex = /.*\/(\w+\/)(\w{40})/
```

Comment:
AFAICS you only need this regex to extract the commit hash so it can be simplified to `.+\/(\w{40})`. Otherwise. if you need to do any kind of validation, I'd rewrite this as a function like so:

```
parseCommitHash = (image) ->
  parts = image?.split('/')
  return null if parts?.length isnt 3 // or throw error
  return parts.pop()
```